docs/custom-transfers.md

```diff
@@ -0,0 +1,276 @@
+
+* `lfs.customtransfer.<name>.concurrent`
+
+  If true (the default), git-lfs will invoke the custom transfer process
+  multiple times in parallel, according to `lfs.concurrenttransfers`, splitting
+  the transfer workload between the processes.
+
+  If you would prefer that only one instance of the transfer process is invoked,
+  maybe because you want to do your own parallelism internally (e.g. slicing 
+  files into parts), set this to false.
+
+* `lfs.customtransfer.<name>.direction`
+
+  Specifies which direction the custom transfer process supports, either 
+  "download", "upload", or "both". The default if unspecified is "both".
+
+## Naming
+
+Each custom transfer must have a name which is unique to the underlying
+mechanism, and the client and the server must agree on that name. The client
+will advertise this name to the server as a supported transfer approach, and if
+the server supports it, it will return relevant object action links. Because
+these may be very different from standard HTTP URLs it's important that the
+client and server agree on the name.
+
+For example, let's say I've implemented a custom transfer process which uses
+NFS. I could call this transfer type 'nfs' - although it's not specific to my
+configuration exactly, it is specific to the way NFS works, and the server will
+need to give me different URLs. Assuming I define my transfer like this, and the
+server supports it, I might start getting object action links back like
+`nfs://<host>/path/to/object`
+
+## Protocol
+
+The git-lfs client communicates with the custom transfer process via the stdin
+and stdout streams. No file content is communicated on these streams, only
+request / response metadata. The metadata exchanged is always in JSON format.
+External files will be referenced when actual content is exchanged.
+
+### Line Delimited JSON 
+Because multiple JSON messages will be exchanged on the same stream it's useful
+to delimit them explicitly rather than have the parser find the closing `}` in
+an arbitrary stream, therefore each JSON structure will be sent and received on
+a **single line** as per [Line Delimited
+JSON](https://en.wikipedia.org/wiki/JSON_Streaming#Line_delimited_JSON_2).
+
+In other words when git-lfs sends a JSON message to the custom transfer it will
+be on a single line, with a line feed at the end. The transfer process must
+respond the same way by writing a JSON structure back to stdout with a single
+line feed at the end (and flush the output).
+
+### Protocol Stages
+
+The protocol consists of 3 stages:
+
+#### Stage 1: Intiation
+
+Immediately after invoking a custom transfer process, git-lfs sends initiation
+data to the process over stdin. This tells the process useful information about
+the configuration.
+
+The message will look like this:
+```json
+{ "id":"init", "operation":"download", "concurrent": true, "concurrenttransfers": 3 }
+```
+
+* `id`: Always "init" to identify this message
```

Comment:
I thought about this far too long as well - naming eh? ;) I deliberately avoided `operation` because of the ambiguity with the batch api requests, and it's already used in the init structure to reflect that very purpose. I originally used `type` but that clashes with reserved words in most languages so I didn't want to cause people with autogen tools more hassle than I needed to.

I'd like `op` if it wasn't a contraction of `operation` which is already used. 😕 

I like `event` best out of the remaining options.